README.md
# Yii2 Toastr

Simple flash toastr notifications for Yii2

[![Latest Stable Version](https://img.shields.io/packagist/v/diecoding/yii2-toastr?label=stable)](https://packagist.org/packages/diecoding/yii2-toastr)
[![Total Downloads](https://img.shields.io/packagist/dt/diecoding/yii2-toastr)](https://packagist.org/packages/diecoding/yii2-toastr)
[![Latest Stable Release Date](https://img.shields.io/github/release-date/sugeng-sulistiyawan/yii2-toastr)](https://github.com/sugeng-sulistiyawan/yii2-toastr)
[![Quality Score](https://img.shields.io/scrutinizer/quality/g/sugeng-sulistiyawan/yii2-toastr)](https://scrutinizer-ci.com/g/sugeng-sulistiyawan/yii2-toastr)
[![Build Status](https://img.shields.io/travis/com/sugeng-sulistiyawan/yii2-toastr)](https://app.travis-ci.com/sugeng-sulistiyawan/yii2-toastr)
[![License](https://img.shields.io/github/license/sugeng-sulistiyawan/yii2-toastr)](https://github.com/sugeng-sulistiyawan/yii2-toastr)
[![PHP Version Require](https://img.shields.io/packagist/dependency-v/diecoding/yii2-toastr/php?color=6f73a6)](https://packagist.org/packages/diecoding/yii2-toastr)

> Yii2 Toastr uses [Toastr](https://codeseven.github.io/toastr/) <br> Demo: https://codeseven.github.io/toastr/demo.html

## Table of Contents

- [Yii2 Toastr](#yii2-toastr)
  - [Table of Contents](#table-of-contents)
  - [Instalation](#instalation)
  - [Dependencies](#dependencies)
  - [Usage](#usage)
    - [Layouts/Views](#layoutsviews)
      - [Layouts/Views Simple Usage](#layoutsviews-simple-usage)
      - [Layouts/Views Advanced Usage](#layoutsviews-advanced-usage)
    - [Controllers](#controllers)
      - [Controllers Simple Usage](#controllers-simple-usage)
      - [Controllers Advanced Usage](#controllers-advanced-usage)

## Instalation

Package is available on [Packagist](https://packagist.org/packages/diecoding/yii2-toastr), you can install it using [Composer](https://getcomposer.org).

```shell
composer require diecoding/yii2-toastr "^1.0"
```

or add to the require section of your `composer.json` file.

```
"diecoding/yii2-toastr": "^1.0"
```

## Dependencies

- PHP 7.4+
- [yiisoft/yii2](https://github.com/yiisoft/yii2)
- [npm-asset/toastr](https://asset-packagist.org/package/npm-asset/toastr)

## Usage

### Layouts/Views

> Add `ToastrFlash` to your layout or view file, example in file `views\layouts\main.php`

#### Layouts/Views Simple Usage

```php
use diecoding\toastr\ToastrFlash;

ToastrFlash::widget();
```

#### Layouts/Views Advanced Usage

```php
use diecoding\toastr\ToastrFlash;

ToastrFlash::widget([
    "typeDefault"       => ToastrFlash::TYPE_INFO,            // (string) default `ToastrFlash::TYPE_INFO`
    "titleDefault"      => "",                                // (string) default `""`
    "messageDefault"    => "",                                // (string) default `""`
    "closeButton"       => false,                             // (bool) default `false`
    "debug"             => false,                             // (bool) default `false`
    "newestOnTop"       => true,                              // (bool) default `true`
    "progressBar"       => true,                              // (bool) default `true`
    "positionClass"     => ToastrFlash::POSITION_TOP_RIGHT,   // (string) default `ToastrFlash::POSITION_TOP_RIGHT`
    "preventDuplicates" => true,                              // (bool) default `true`
    "showDuration"      => 300,                               // (int|null) default `300` in `ms`, `null` for skip
    "hideDuration"      => 1000,                              // (int|null) default `1000` in `ms`, `null` for skip
    "timeOut"           => 5000,                              // (int|null) default `5000` in `ms`, `null` for skip
    "extendedTimeOut"   => 1000,                              // (int|null) default `1000` in `ms`, `null` for skip
    "showEasing"        => "swing",                           // (string) default `swing`, `swing` and `linear` are built into jQuery
    "hideEasing"        => "swing",                           // (string) default `swing`, `swing` and `linear` are built into jQuery
    "showMethod"        => "slideDown",                       // (string) default `slideDown`, `fadeIn`, `slideDown`, and `show` are built into jQuery
    "hideMethod"        => "slideUp",                         // (string) default `slideUp`, `hide`, `fadeOut` and `slideUp` are built into jQuery
    "tapToDismiss"      => true,                              // (bool) default `true`
    "escapeHtml"        => true,                              // (bool) default `true`
    "rtl"               => false,                             // (bool) default `false`
    "skipCoreAssets"    => false,                             // (bool) default `false`, `true` if use custom or external toastr assets
    "options"           => [],                                // (array) default `[]`, Custom Toastr options and override default options
]);
```

### Controllers

> Just use `Yii::$app->session->setFlash($type, $message)` like as usual alert
#### Controllers Simple Usage

```php
Yii::$app->session->setFlash('error', 'Message');
```

or if use multiple flash in same session

```php
Yii::$app->session->setFlash('error', ['Message 1', 'Message 2', 'Message 3']);
```

#### Controllers Advanced Usage

```php
Yii::$app->session->setFlash('error', [['Title', 'Message']]);
```

or if use multiple flash in same session

```php
Yii::$app->session->setFlash('error', [['Title 1', 'Message 1'], ['Title 2', 'Message 2'], ['Title 3', 'Message 3']]);
```


CHANGELOG.md
# Changelog

All Notable changes to `yii2-toastr` will be documented in this file.

## 2.0.0 - March 30, 2023
Major Update

## Changed
- composer
- phpunit
- namespace
- license
- and more

### Added
- tests

### Deprecated
- namespace
- php 7.0

### Fixed
- Nothing

### Removed
- php 7.0

### Security
- Nothing

## 1.0.4 - August 12, 2019

## Changed
- phpunit

### Added
- Nothing

### Deprecated
- Nothing

### Fixed
- bug

### Removed
- travis

### Security
- Nothing

---

## 1.0.3 - Ocktober 22, 2018

## Changed
- src\\Toastr.php
- src\\ToastrBase.php
- src\\ToastrFlash.php

### Added
- tests

### Deprecated
- Nothing

### Fixed
- bug style codes

### Removed
- Nothing

### Security
- Nothing

---

## 1.0.2 - Ocktober 21, 2018

## Changed
- src\\ToastrBase.php
- src\\ToastrFlash.php

### Added
- CHANGELOG.md

### Deprecated
- Nothing

### Fixed
- Nothing

### Removed
- Nothing

### Security
- Nothing

---

## 1.0.1 - September 25, 2018

## Changed
- Nothing

### Added
- README.md

### Deprecated
- Nothing

### Fixed
- Nothing

### Removed
- Nothing

### Security
- Nothing

---

## 1.0.0 - September 25, 2018

## Changed
- Nothing

### Added
- Nothing

### Deprecated
- Nothing

### Fixed
- Nothing

### Removed
- Nothing

### Security
- Nothing


src/ToastrBase.php
<?php

namespace diecoding\toastr;

use yii\base\Widget;
use yii\helpers\ArrayHelper;

/**
 * ToastrBase is the base class for widgets.
 * 
 * @link [sugeng-sulistiyawan.github.io](sugeng-sulistiyawan.github.io)
 * @author Sugeng Sulistiyawan <sugeng.sulistiyawan@gmail.com>
 * @copyright Copyright (c) 2023
 */
class ToastrBase extends Widget
{
	const TYPE_INFO    = "info";
	const TYPE_ERROR   = "error";
	const TYPE_SUCCESS = "success";
	const TYPE_WARNING = "warning";

	const POSITION_TOP_RIGHT      = "toast-top-right";
	const POSITION_TOP_LEFT       = "toast-top-left";
	const POSITION_TOP_CENTER     = "toast-top-center";
	const POSITION_TOP_FULL_WIDTH = "toast-top-full-width";

	const POSITION_BOTTOM_RIGHT      = "toast-bottom-right";
	const POSITION_BOTTOM_LEFT       = "toast-bottom-left";
	const POSITION_BOTTOM_CENTER     = "toast-bottom-center";
	const POSITION_BOTTOM_FULL_WIDTH = "toast-bottom-full-width";

	const TYPES = [
		self::TYPE_INFO,
		self::TYPE_ERROR,
		self::TYPE_SUCCESS,
		self::TYPE_WARNING,
	];

	const POSITIONS = [
		self::POSITION_TOP_RIGHT,
		self::POSITION_TOP_LEFT,
		self::POSITION_TOP_CENTER,
		self::POSITION_TOP_FULL_WIDTH,

		self::POSITION_BOTTOM_RIGHT,
		self::POSITION_BOTTOM_LEFT,
		self::POSITION_BOTTOM_CENTER,
		self::POSITION_BOTTOM_FULL_WIDTH,
	];

	/**
	 * @var string default `self::TYPE_INFO`
	 */
	public $typeDefault = self::TYPE_INFO;

	/**
	 * @var string default `""`
	 */
	public $titleDefault = "";

	/**
	 * @var string default `""`
	 */
	public $messageDefault = "";

	/**
	 * @var bool default `false`
	 */
	public $closeButton = false;

	/**
	 * @var bool default `false`
	 */
	public $debug = false;

	/**
	 * @var bool default `true`
	 */
	public $newestOnTop = true;

	/**
	 * @var bool default `true`
	 */
	public $progressBar = true;

	/**
	 * @var string default `self::POSITION_TOP_RIGHT`
	 */
	public $positionClass = self::POSITION_TOP_RIGHT;

	/**
	 * @var bool default `true`
	 */
	public $preventDuplicates = true;

	/**
	 * @var int|null default `300` in `ms`, `null` for skip
	 */
	public $showDuration = 300;

	/**
	 * @var int|null default `1000` in `ms`, `null` for skip
	 */
	public $hideDuration = 1000;

	/**
	 * @var int|null default `5000` in `ms`, `null` for skip
	 */
	public $timeOut = 5000;

	/**
	 * @var int|null default `1000` in `ms`, `null` for skip
	 */
	public $extendedTimeOut = 1000;

	/**
	 * @var string default `swing`, `swing` and `linear` are built into jQuery
	 */
	public $showEasing = "swing";

	/**
	 * @var string default `swing`, `swing` and `linear` are built into jQuery
	 */
	public $hideEasing = "swing";

	/**
	 * @var string default `slideDown`, `fadeIn`, `slideDown`, and `show` are built into jQuery
	 */
	public $showMethod = "slideDown";

	/**
	 * @var string default `slideUp`, `hide`, `fadeOut` and `slideUp` are built into jQuery
	 */
	public $hideMethod = "slideUp";

	/**
	 * @var bool default `true`
	 */
	public $tapToDismiss = true;

	/**
	 * @var bool default `true`
	 */
	public $escapeHtml = true;

	/**
	 * @var bool default `false`
	 */
	public $rtl = false;

	/**
	 * @var bool default `false`, `true` if use custom or external toastr assets
	 */
	public $skipCoreAssets = false;

	/**
	 * @var array default `[]`, Custom Toastr options and override default options
	 */
	public $options = [];

	/**
	 * @inheritdoc
	 */
	public function init()
	{
		if ($this->skipCoreAssets === false) {
			$this->view->registerAssetBundle(ToastrAsset::class);
		}

		$this->options = ArrayHelper::merge([
			"closeButton"       => $this->closeButton,
			"debug"             => $this->debug,
			"newestOnTop"       => $this->newestOnTop,
			"progressBar"       => $this->progressBar,
			"positionClass"     => $this->positionClass,
			"preventDuplicates" => $this->preventDuplicates,
			"showDuration"      => $this->showDuration,
			"hideDuration"      => $this->hideDuration,
			"timeOut"           => $this->timeOut,
			"extendedTimeOut"   => $this->extendedTimeOut,
			"showEasing"        => $this->showEasing,
			"hideEasing"        => $this->hideEasing,
			"showMethod"        => $this->showMethod,
			"hideMethod"        => $this->hideMethod,
			"tapToDismiss"      => $this->tapToDismiss,
			"escapeHtml"        => $this->escapeHtml,
			"rtl"               => $this->rtl,
		], $this->options);

		parent::init();
	}
}
